### DISCUSSION TITLE
presence simulation with timer rule and random

### DISCUSSION THREAD

**Discussion by bernd208** [arendst/Tasmota#20571] [Q&A]
Hardware nous A1T
SW Version 13.3.0
Question:
I am trying to program a presence simulation here. The 16 timers can be started with a random deviation of +-15 minutes and act as an action on a rule. In rule1, I can set a separate delay time for each timer1-16:

12:46:23.793 CMD: rule1 on Clock#Timer=5 do backlog power on; ruletimer1 5 endon on rules#timer=1 do power off endon
12:46:23.798 RUL: Stored uncompressed, would compress from 92 to 64 (-30%)
12:46:23.801 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":92,"Free":419,"Rules":"on Clock#Timer=5 do backlog power on; ruletimer1 5 endon on rules#timer=1 do power off endon"}}
12:46:56.399 CMD: #
12:47:00.431 RUL: CLOCK#TIMER=5 performs "backlog power on; ruletimer1 5"
12:47:00.464 RSL: RESULT = {"POWER":"ON"}
12:47:00.466 RSL: POWER = ON
12:47:00.716 RSL: RESULT = {"T1":5,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
12:47:06.455 RUL: RULES#TIMER=1 performs "power off"
12:47:06.460 RSL: RESULT = {"POWER":"OFF"}
12:47:06.462 RSL: POWER = OFF
T1:5 was set in the result. And power was turned off after 5 seconds. It works.

Now I also want to randomly determine the delay time in the range of [x, y] seconds:

12:50:58.126 CMD: rule1 on Clock#Timer=5 do backlog power on; ruletimer1 [random(3, 5)] endon on rules#timer=1 do power off endon
12:50:58.133 RUL: Stored uncompressed, would compress from 105 to 78 (-26%)
12:50:58.136 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":105,"Free":406,"Rules":"on Clock#Timer=5 do backlog power on; ruletimer1 [random(3, 5)] endon on rules#timer=1 do power off endon"}}
12:51:00.441 RUL: CLOCK#TIMER=5 performs "backlog power on; ruletimer1 [random(3, 5)]"
12:51:00.474 RSL: RESULT = {"POWER":"ON"}
12:51:00.476 RSL: POWER = ON
12:51:00.701 RSL: RESULT = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
Power was turned on, but T1was set to 0.
Unfortunately, the relay is only turned on and not off. Suggestion from chatgpt: move random in the rule to the front:
12:56:43.883 CMD: #on Clock#Timer=5 do var %randomTime = random(3, 5) backlog power on; ruletimer1 %randomTime endon on rules#timer=1 do power off endon

12:57:00.456 RUL: CLOCK#TIMER=5 performs "backlog power on; ruletimer1 [random(3, 5)]"
12:57:00.488 RSL: RESULT = {"POWER":"ON"}
12:57:00.490 RSL: POWER = ON
12:57:00.691 RSL: RESULT = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}

T1:0 was not set here either.
3rd attempt:
13:02:30.546 CMD: rule1 on Clock#Timer=5 do var %randomTime = random(3, 5) backlog power on; ruletimer1 %randomTime endon on rules#timer=1 do power off endon
13:02:30.554 RUL: Stored uncompressed, would compress from 139 to 91 (-34%)
13:02:30.557 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":139,"Free":372,"Rules":"on Clock#Timer=5 do var %randomTime = random(3, 5) backlog power on; ruletimer1 %randomTime endon on rules#timer=1 do power off endon"}}
13:03:00.460 RUL: CLOCK#TIMER=5 performs "var %randomTime = random(3, 5) backlog power on; ruletimer1 %randomTime"
13:03:00.467 RSL: RESULT = {"Var1":"","Var2":"","Var3":"","Var4":"","Var5":"","Var6":"","Var7":"","Var8":"","Var9":"","Var10":"","Var11":"","Var12":"","Var13":"","Var14":"","Var15":"","Var16":""}
This also doesn't work.

Question: Does random work for the delay time in the rule? Or what am I doing wrong?

**Comment by sfromis** [comment]
Have you been asking ChatGPT for help? It looks like the type of bonkers hallucination answer you'd get from there. Rules have no feature for random mumbers.

**Reply by sfromis** [reply]
In some devices, you may be able to pick up some unpredictable numbers by configuring analog on gpio 17 (normally not connected), meaning that random noise can affect what numbers you get.

**Reply by bernd208** [reply]
ok,
the template is: {"NAME":"NOUS A1T","GPIO":[32,0,0,0,2720,2656,0,0,2624,320,224,0,0,4704],"FLAG":0,"BASE":49}
i set the gpio17 to ADCInput 
and in the console: 
AnalogRead 17  or A0 or
AnalogRead(17)  or (A0)
but get no value
whats wrong?

**Reply by sfromis** [reply]
You did not post the context of what you're mentioning there, and I do not recognize it as anything.
What I get at TelePeriod is payloads like:
`{"Time":"2024-01-23T18:45:24.733+01:00","ANALOG":{"A0":179}}`
And on the Tasmota front page: 
![image](https://github.com/arendst/Tasmota/assets/47082390/88967cde-f8da-4906-a908-be49a7f5fa2c)

With the values varying between about 150 and 240. However, this may depend on details of the board design.

**Reply by bernd208** [reply]
the plug needs a reset...

**Reply by bernd208** [reply]
How can i read the value in cmd?

**Reply by sfromis** [reply]
You can't "read" it. What you can do instead is to have a rule reacting to the value being probed. Something like `on ANALOG#A0 do var1 %value% endon` or only at TelePeriod: `on tele-ANALOG#A0 do var1 %value% endon`

**Reply by bernd208** [reply]
thanks, now i stop searching for it...

**Reply by bernd208** [reply]
it works fine:
Rule1 on analog#A0   do var1 %value%; EndOn   on Clock#Timer=5 do backlog power on; ruletimer1 %var1% endon on rules#timer=1 do power off endon
thanks

**Comment by bernd208** [comment]
The only alternative for presence simulation that I have found so far is to set all the timers to toggle (without using rules) with a random deviation of ±15 minutes and set the Spacing between timers to 18 minutes. This creates a potentially chaotic on-off pattern during the evening hours.

**Reply by barbudor** [reply]
Scripting Language has some options for random numbers and may be an alternative to Rules more suited to your use case

**Reply by bernd208** [reply]
ah, ok, i found : 
https://tasmota.github.io/docs/Berry-Introduction/
make a tea and ....
Thanks

**Reply by barbudor** [reply]
No
Berry is only for ESP32

Scripting-Language is here: https://tasmota.github.io/docs/Scripting-Language/

**Reply by bernd208** [reply]
ok, thanks...

**Reply by bernd208** [reply]
I have now realized that in version 13.3.0, I do not have the 'script edit' option. I also do not want to switch to another version and am content with what is available.
Thanks

**Comment by bernd208** [comment]
I have reset and reconfigured the Nous. After setting GPIO17 to ADCInput, the value is displayed on the main page. I will now try to use this as a random input to simulate presence through lamps.
Thanks and Kind Regards
